rqd/rqd/rqnimby.py

```diff
@@ -64,114 +76,210 @@ def lockNimby(self):
                                           self.lockedInUse)
             self.thread.start()
 
     def lockedIdle(self):
         """Nimby State: Machine is idle,
                         waiting for sufficient idle time to unlock"""
-        self._openEvents()
+        self.openEvents()
         waitStartTime = time.time()
         try:
             self.results = select.select(self.fileObjList, [], [],
                                          rqd.rqconstants.MINIMUM_IDLE)
         except:
             pass
         if self.active and self.results[0] == [] and \
-           self.rqCore.machine.isNimbySafeToUnlock():
-            self._closeEvents()
+                self.rqCore.machine.isNimbySafeToUnlock():
+            self.closeEvents()
             self.unlockNimby(asOf=waitStartTime)
             self.unlockedIdle()
         elif self.active:
-            self._closeEvents()
+            self.closeEvents()
+            self.thread = threading.Timer(rqd.rqconstants.CHECK_INTERVAL_LOCKED,
+                                          self.lockedInUse)
+            self.thread.start()
+
+    def openEvents(self):
+        """Opens the /dev/input/event* files so nimby can monitor them"""
+        self.closeEvents()
+
+        rqd.rqutil.permissionsHigh()
+        try:
+            for device in os.listdir("/dev/input/"):
+                if device.startswith("event") or device.startswith("mice"):
+                    log.debug("Found device: %s" % device)
+                    try:
+                        self.fileObjList.append(open("/dev/input/%s" % device, "rb"))
+                    except IOError as e:
+                        # Bad device found
+                        log.debug("IOError: Failed to open %s, %s" % ("/dev/input/%s" % device, e))
+        finally:
+            rqd.rqutil.permissionsLow()
+
+    def closeEvents(self):
+        """Closes the /dev/input/event* files"""
+        log.debug("closeEvents")
+        if self.fileObjList:
+            for fileObj in self.fileObjList:
+                try:
+                    fileObj.close()
+                except:
+                    pass
+            self.fileObjList = []
+
+
+class NimbyPynput(Nimby):
+    def __init__(self, rqCore):
+        Nimby.__init__(self, rqCore)
+
+        import pynput
+        self.mouse_listener = pynput.mouse.Listener(
+            on_move=self.on_interaction,
+            on_click=self.on_interaction,
+            on_scroll=self.on_interaction)
+        self.keyboard_listener = pynput.keyboard.Listener(on_press=self.on_interaction)
+
+    def on_interaction(self, *args):
+        self.interaction_detected = True
+
+    def startListener(self):
+        self.mouse_listener.start()
+        self.keyboard_listener.start()
+
+    def stopListener(self):
+        self.mouse_listener.stop()
+        self.keyboard_listener.stop()
+
+    def lockedInUse(self):
+        self.interaction_detected = False
+
+        time.sleep(5)
```

Comment:
What are the `sleep`s doing here? Waiting for some service to come up or something like that? If so, anything more reliable we can use?

Reply:
This part of the code was suggested by #613 . I assume this is waiting for an interaction to be detected. 

I've added a new constant for this.

Reply:
Ok, I see. It feels like we're double-sleeping here though?

For example `lockedInUse` in this PR has two wait periods built in:
1. First it `sleep`s to wait for interaction.
1. It checks if interaction was detected, if it was not is goes into `lockedIdle` state.
1. If no interaction was detected, it starts `threading.Timer` to check again in `CHECK_INTERVAL_LOCKED` seconds.

So it seems like `sleep` is duplicating the `Timer`. Am I understanding this correctly?

The `sleep` has other minor side effects like making testing more difficult so I want to make sure it's actually needed.